### PR TITLE
Implement Unpin for all raw pointers

### PR DESCRIPTION
**Pull request by sfackler**
Like references and boxes, moving the pointer doesn't move the
pointed-to value, so this is safe.

r? @cramertj

### CODE REVIEW

**Comment by cramertj** [comment]
cc @taiki-e , @withoutboats , @Nemo157 

I don't have a particularly strong opinion here. We left this out initially as a conscious decision since we weren't sure if there were types in the wild that might want to do `Unpin` projection through raw pointers, such as I initially proposed we do with `Box`. We wound up going the other way on `Box`, primarily due to the simpler mental model of `Pin`-projection as something that happens inwards in memory, rather than traveling through pointers. I still mildly feel like that was a mistake since it meant that `Box<dyn Future>` doesn't implement `Future`, but the decision is past us now and it seems unlikely that we'd want a third-party crate to go a different direction.

TL;DR: LGTM.

**Comment by ProgrammaticNajel** [comment]
Ping from triage @cramertj any updates on this?

**Comment by Centril** [comment]
cc @RalfJung also.

**Comment by RalfJung** [comment]
LGTM.

> it meant that Box<dyn Future> doesn't implement Future

`Pin<Box<dyn Future>>` should implement `Future` though, right?

**Comment by Nemo157** [comment]
> `Pin<Box<dyn Future>>` should implement `Future` though, right?

Yes, `Pin<impl DerefMut<impl Future + ?Sized>>` and `Box<impl Future + Unpin + ?Sized>` both implement `Future`.
